File: Pre CAT/DS/Stack.c
#include<stdio.h>
int isEmpty(int *top){
    if(*top == -1)
        return 1;
    else
        return 0;
}
int isFull(int *top,int size){
    if(*top==size-1){
        return 1;
    }else{
        return 0;
    }
}

void push(int *arr,int *top,int size,int data){
    if(!isFull(top,size)){
        arr[++(*top)]=data;
    }else{
        printf("Stack is full\n");
    }
}

void pop(int* arr,int* top){
    if(!isEmpty(top)){
        printf("Poped element: %d\n",arr[(*top)]);
        (*top)--;
    }else{
        printf("Stack is empty\n");
    }
}

void display(int* arr,int* top){
    if(!isEmpty(top)){
        for(int i=0;i<=(*top);i++)
            printf("%d ",arr[i]);
        printf("\n");
    }else{
        printf("Stack is empty\n");
    }
           
}

void elementAtTop(int* arr, int* top){
    if(!isEmpty(top))
        printf("Element at top: %d\n",arr[(*top)]);
    else
        printf("Stack is empty\n");
}
int main(){
    int size;
    printf("Enter the size of stack: ");
    scanf("%d",&size);
    int arr[size];
    int top = -1;
    push(arr,&top,size,5);
    push(arr,&top,size,10);
    display(arr,&top);
    push(arr,&top,size,15);
    push(arr,&top,size,20);
    display(arr,&top);
    pop(arr,&top);
    display(arr,&top);
    elementAtTop(arr,&top);
    push(arr,&top,size,25);
    display(arr,&top);
    pop(arr,&top);
    pop(arr,&top);
    pop(arr,&top);
    pop(arr,&top);
    display(arr,&top);
    elementAtTop(arr,&top);
}

File: CDAC/DSAUsingJava/src/dataStructure/Stack.java
package dataStructure;

import java.util.Scanner;

public class Stack {
	
	private int[] arr;
	private int top;
	private int size;
	
	public Stack(int size) {
		this.size = size;
		this.arr = new int[size];
		this.top = -1;
	}
	
	boolean isEmpty() {
		if(top == -1)
			return true;
		else
			return false;
	}
	
	boolean isFull() {
		if(top == size-1)
			return true;
		else
			return false;
	}
	
	void push(int data) throws Exception {
		if(!isFull()) {
			arr[++top] = data;
			System.out.println(data+" inserted successfully!");
		}else
			throw new Exception("Stack is full!!");
	}
	
	int pop() throws Exception{
		if(!isEmpty())
			return arr[top--];
		else
			throw new Exception("Stack is empty!!");
	}
	
	int peek() throws Exception{
		if(!isEmpty())
			return arr[top];
		else
			throw new Exception("Stack is empty!!");
	}
	
	void display() throws Exception{
		if(!isEmpty()) {
			for(int i=0;i<=top;i++)
				System.out.print(arr[i]+" ");
			System.out.println();
		}else
			throw new Exception("Stack is empty!!");
	}
	
	public static void main(String[] args) {
		System.out.print("Enter the size of the stack:");
		Scanner sc = new Scanner(System.in);
		int size,choice,data,element;
		size = sc.nextInt();
		Stack s = new Stack(size);
		while(true) {
			System.out.println("1. Push");
			System.out.println("2. Pop");
			System.out.println("3. Peek");
			System.out.println("4. Display");
			System.out.println("5. Exit");
			System.out.print("Enter your choice:");
			choice = sc.nextInt();
			try {
				switch(choice) {
				case 1:
					System.out.println("Enter the data to be pushed");
					data = sc.nextInt();
					s.push(data);
					break;
				case 2:
					element = s.pop();
					System.out.println("Poped element is "+element);
					break;
				case 3:
					element = s.peek();
					System.out.println("Element at top is "+element);
					break;
				case 4:
					s.display();
					break;
				case 5:
					System.exit(0);
				}
				
			} catch (Exception e) {
				System.out.println(e.getMessage());
			}
			
		}
	}
}
